Disabling sync and standalone features.

jankrzempek opened the pull request:
As a result of situations where users were not able to use those features properly - disabling them from now for iPhone SE 1st GEN is the solution.

**What is the scope of this PR:**

- Disable timeout for  BT reconnection feature when using iPhone SE. Instead, show the new card view and wait until AB is reconnected.
- Disable the ability to `finish and sync` when using iPhone SE. Instead, show the card for infinite reconnection

jankrzempek committed:
Disabling sync and standalone features.

As the result of situations where users were not able to use those features properly - those will be disabled from now for iPhone SE 1st GEN.

anna1901 reviewed (changes requested):
I checked on the iPhone SE 1st gen simulator and SD sync button is present. Does Device.current work also with simulators?

jankrzempek committed:
Added simulators to excluded deviced and ranamed func.

anna1901 reviewed (changes requested):
When the AB disconnects from the phone we are automatically putting session in standalone mode. I think this situation should also be handled somehow and we should show some other card than for standalone sessions (because there is this finish and sync button on those cards)

anna1901 commented:
![simulator_screenshot_CC9ED068-A4B8-442F-BA89-067AF4456C7E](https://user-images.githubusercontent.com/15946812/196165043-fdd7e090-0d74-493b-8e2d-561265e138b2.png)
I this the look we are going for? 😄

jankrzempek committed:
Merge branch 'develop' into restrict_sd_sync_standalone

jankrzempek committed:
Infinite reconnection for iPhone SE and 60s for others.

anna1901 reviewed (changes requested):
I added my device to excluded devices list and I can still see sync button and can conduct sync

anna1901 commented on AirCasting/SessionViews/ReconnectSessionCardView.swift:
```diff
@@ -0,0 +1,101 @@
+        .padding()
+    }
+    
+    var reconnectionLabel: some View {
```
In my opinion this design should be rethinked because blue button suggest that it is tappable and some action will follow while it is not

jankrzempek replied:
This is a design provided to me by Michael (it comes from Android) - and I don't think it is worth the effort to now change this design in both systems. What do you think?

![Screenshot_20221017-161911](https://user-images.githubusercontent.com/44276070/197533666-f37385bb-3771-44bf-bd00-1fbfbbe8bafb.png)

jankrzempek replied:
I will ask Michael, never thought about this

jankrzempek commented on AirCasting/ABConnector/BluetoothManager.swift:
```diff
@@ -211,9 +211,13 @@ extension BluetoothManager: CBCentralManagerDelegate {
         guard mobilePeripheralSessionManager.activeSessionInProgressWith(peripheral) else { return }
         mobilePeripheralSessionManager.markActiveSessionAsDisconnected(peripheral: peripheral)
         connect(to: peripheral)
```
Yes, we are then trying to connect unless we will. Infinite reconnection should be applied to iPhone SE (info in the PR description).

jankrzempek committed:
Conditionall sdSync button and improvements.

jankrzempek commented:
> I added my device to excluded devices list and I can still see sync button and can conduct sync

Thanks so much, I forgot that I removed it when doing testing 🫡

pawelgil commented on AirCasting.xcodeproj/project.pbxproj:
```diff
@@ -2724,6 +2736,16 @@
+			isa = PBXGroup;
+			children = (
+				FA2438F328F996820028B126 /* Features.swift */,
+				FA2438F028F990260028B126 /* DeviceKit+checkDeviceSupportFor.swift */,
+				FA2438F528F996A00028B126 /* Device+excluded.swift */,
```
extension names are capitalized by convention:

`Device+excluded.swift -> Device+Excluded.swift`
`DeviceKit+check... -> DeviceKit+Check...`

pawelgil commented on AirCasting/SessionViews/ReconnectSessionCardView.swift:
```diff
@@ -0,0 +1,95 @@
+import Resolver
+
+struct ReconnectSessionCardView: View {
+    let session: SessionEntity
```
do we really need this here? Maybe we can use `Sessionable` instead? We're trying not to use CoreData entities in views anymore

pawelgil commented on AirCasting/SessionViews/ReconnectSessionCardView.swift:
```diff
@@ -0,0 +1,95 @@
+        .font(Fonts.moderateRegularHeading2)
+    }
+    
+    func finishSessionAlertAction() {
```
can we create a proper ViewModel for this view and move business logic there?

pawelgil commented on AirCasting/SessionViews/SessionHeaderView.swift:
```diff
@@ -289,3 +289,14 @@ private extension SessionHeaderView {
+        if session.deviceType == .AIRBEAM3 &&
+            session.isActive && featureFlagsViewModel.enabledFeatures.contains(.standaloneMode) &&
+            checkDeviceSupportFor(feature: .standalone) {
+            return true
```
we can shorten it from
```
if condition {
  return true
} else {
  return false
}
```
to just:
```
return condition
```

pawelgil reviewed (changes requested):
I'd go a different route here.

### Why
This `protocol FeatureProtocol` and `enum Features`  can and will be easily confused with our feature flags. It's adding a lot of complexity.

### Observation
I think this doesn't really differ from our "standard" features controlled by feature flags.

### Idea
Instead of introducing a whole separate mechanism for those new features and spilling the new `DeviceKit` dependency all over the project files maybe it would be cleaner to fit this into our existing feature flagging code? We already have all the machinery at its core including compositional helpers (composite class). I think this design would be good:
current feature flags setup (for `RELEASE` builds):
![flags_before](https://user-images.githubusercontent.com/10663361/198366802-331b91d0-6a90-442c-8757-4bab2c5d32f5.jpg)

to this we just add a new provider (note how nicely the open/closed principle works here):
![flags_new](https://user-images.githubusercontent.com/10663361/198367351-13e3dfd6-b362-43c2-a43b-74bbb4326008.jpg)

------
### Implementation
* This new provider should be the most important one - if the feature is not available (`on`) on this device type we don't care if it's enabled on firebase
* There already is a feature flag for standalone mode and sd sync - will it be ok to use those or do we need new ones? (I think we need new ones)
* Make sure we don't leak `DeviceKit` outside of the provider impl
* Wherever needed just `@Inject private var featureFlags: FeatureFlagProvider` and then `featureFlags.isFeatureOn(...)` instead of the current `Device` based code.

EDIT: I know I'm late to the party and dropping this comment, sorry for that @jankrzempek

jankrzempek committed:
BT reconnection apllied along with DeviceFeatureFlag.

anna1901 commented on AirCasting/ABConnector/BluetoothManager.swift:
```diff
@@ -49,9 +49,12 @@ class BluetoothManager: NSObject, BluetoothCommunicator, ObservableObject {
     @Published var mobileSessionReconnected = false
     var observed: NSKeyValueObservation?
+    @Published var isReconnectionOn = false
+    @Published var didAlreayCancellConnection = false
```
I'm not sure about the logic with this variable, because it is used in different functions with the assumption of a specific flow, but okay. I don't like the name of this variable. What do you think about `connectingTimedOut`?

jankrzempek replied:
Typo is fixed, thanks. In addition to the name - I think it appropriately says what it does and what is its purpose. If you will force the change of name I will do it, but another way I would like it to stay

anna1901 replied:
Okay, I will not force anything, but I have to say I don't understand the name. Who already cancelled connection? User? Device?

jankrzempek replied:
That's bad, I will come up with something new and place it here so, please let me know then what you think

jankrzempek replied:
What would you say for `peripheralConnectionIsCancelled`?

anna1901 reviewed (changes requested):
I had a session in standalone mode. Then I excluded standalone feature for iphoneX instead of iphoneSE. I tapped Reconnect. I get that it shouldn't work, but I also didn't get any alert either. The spinner was spinning and then it wasn't.

anna1901 reviewed (changes requested):
Additionally, again, I can see the sd sync button even if I exchange iPhone SE to my device in DeviceFeatureFlagsProvider

jankrzempek commented:
> Additionally, again, I can see the sd sync button even if I exchange iPhone SE to my device in DeviceFeatureFlagsProvider

Checked. After setting `Device.current` to my iPhone, the SD Sync button disappeared nicely. Could tell me more about your case so I can try to reproduce?

jankrzempek committed:
Naming improvements and explanations by comments.

jankrzempek committed:
Changing conditional implementation to enum.

anna1901 reviewed (changes requested):
This doesn't work well. I cannot reconnect with a device. I get an alert that app no longer can connect with a device. I finish session. Then I try to record another session with this AB, I get a connection error, but the previous session that was already finished is revived. I think it is best if you remove the whole reconnecting functionality from this PR and leave just the view (perhaps with the reconnect button commented out) and we will implement the reconnection logic once the bt refactor is done. This implementation is just too buggy for my taste and there is no point merging it to develop branch. But all of the other changes are good, so I think it's best to merge them without the reconnection logic. :)

jankrzempek committed:
Removed logic responsible for reconnection.

jankrzempek committed:
Leftovers.

jankrzempek committed:
More leftovers.

anna1901 commented on AirCasting/ABConnector/BluetoothManager.swift:
```diff
@@ -210,6 +210,10 @@ extension BluetoothManager: CBCentralManagerDelegate {
 
         guard mobilePeripheralSessionManager.activeSessionInProgressWith(peripheral) else { return }
         mobilePeripheralSessionManager.markActiveSessionAsDisconnected(peripheral: peripheral)
+        connectWithTimeout(using: peripheral)
```
Please remove those changes so that there are no conflicts with BT manager (this class is getting deleted anyway)

anna1901 commented on AirCasting/SessionViews/ReconnectSessionCardView.swift:
```diff
@@ -0,0 +1,88 @@
+    var reconnectionLabel: some View {
+        Button {
+            if !viewModel.isSpinnerOn {
+                // Custom method of button disable: this can keep the UI style always the same
```
What do you mean by this comment?

jankrzempek replied:
That we are not using the `.disable` property instead we have let's say custom disable method, which can keep the blue color of the button all the time.

anna1901 commented on AirCasting/SessionViews/ReconnectSessionCardViewModel.swift:
```diff
@@ -0,0 +1,43 @@
+            try sessionStopper.stopSession()
+            completion()
+        } catch {
+            Log.info("error when stpoing session - \(error)")
```
Alert here?

jankrzempek replied:
could be, why not 😅

anna1901 commented on AppDelegate+Injection.swift:
```diff
@@ -111,23 +111,27 @@ extension Resolver: ResolverRegistering {
 #if DEBUG
             CompositeFeatureFlagProvider(children: [
                 Resolver.resolve(OverridingFeatureFlagProvider.self),
+//                Resolver.resolve(DeviceFeatureFlagProvider.self),
```
Why is it commented out?

jankrzempek replied:
To be sure, it won't work unless the whole feature will be implemented.

anna1901 reviewed (changes requested):
Looks quite good, I just added a few comments. I'm gonna test later, so this CR is based only on looking at the code changes

jankrzempek commented on AirCasting/SessionViews/ReconnectSessionCardViewModel.swift:
```diff
@@ -0,0 +1,43 @@
+    let session: SessionEntity
+    @Published var alert: AlertInfo?
+    @Published var isSpinnerOn = false
+    @Injected private var bm: BluetoothManager
```
The idea for this was to stay, so it will be easier to inspire and implement when I will be doing this soon.

jankrzempek committed:
Improvements.

jankrzempek committed:
Improvements.

jankrzempek committed:
typo in error message.

anna1901 reviewed (approved):
No you are not showing that Reconnection card at all, so I'm not sure what I can test here. Maybe implement also the logic for deciding what session card should be shown and then we can merge it? I'm giving the approve, cause I'm leaving this decision to you
